File: controllers/AuthController.php
<?php
class AuthController extends Controller
{
    protected function init()
    {
        App::loadModel('User');
        Session::init();
    }

    // POST /auth/login + JSON Data
    public function login()
    {
        $data = null;
        if (!$this->_isPOSTandHasData($data) ||
            !isset($data['username']) || !isset($data['password']) ||
            Session::loggedIn()) {
            return $this->response->nullJson();
        }

        if (!$user = User::getOneByUsername($data['username'])) {
            return $this->response->nullJson();
        }

        if (md5($data['password']) === $user->get_password()) {
            Session::login($user->get_username());
            return $this->response->renderJson(array('status' => 'success'));
        }

        return $this->response->nullJson();
    }

    // GET /auth/logout
    public function logout()
    {
        if (Session::loggedIn()) {
            Session::logout();
            return $this->response->renderJson(array('status' => 'success'));
        }

        return $this->response->nullJson();
    }
}

File: src/madtodo/controllers/AuthController.java
package madtodo.controllers;

import madtodo.MadController;
import madtodo.models.User;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONTokener;

public class AuthController extends MadController {
    ////
    // url { /auth }
    @Override
    public void index() {
        print404();
    }

    ////
    // url { /auth/login }
    public void login() {
        if (!isCheckPass()) {
            printFailed();
        } else {
            String rawData = getQueryParam("data");

            JSONObject data = null;
            try {
                data = new JSONObject(new JSONTokener(rawData));
            } catch (JSONException e) {
                printJSON(new JSONObject().put("error", "Invalid JSON Data"));
                return;
            }

            String username = data.getString("username");
            String password = data.getString("password");

            User user = User.findByUsername(username);
            if (user == null) {
                printFailed();
                System.out.println("User not found");
            } else {
                if (user.getPassword().equals(md5sum(password))) {
                    session.login(username, user.getId());
                    printSuccess();
                } else {
                    printFailed();
                    System.out.println("Wrong password");
                }
            }
        }
    }

    ////
    // url { /auth/logout }
    public void logout() {
        if (session.isLoggedIn()) {
            session.logout();
            printSuccess();
        } else {
            printFailed();
        }
    }

    private boolean isCheckPass() {
        boolean pass = true;

        if (request.getMethod() != "POST") {
            pass = false;
        }

        if (request.getParameter("data") == null) {
            pass = false;
        }

        return pass;
    }
}
